compiler/il/OMRNode.cpp

```diff
@@ -2316,17 +2316,115 @@ OMR::Node::dontEliminateStores(bool isForLocalDeadStore)
-      return true;
-   else if (comp->generateArraylets() &&
-            (self()->getOpCode().isArrayRef() &&
-             (self()->getFirstChild()->isNotCollected() || !self()->getFirstChild()->isInternalPointer())))
-      return true;
-   return false;
+   return !self()->computeIsCollectedReference();
    }
 
 
+/**
+ * \brief
+ *    Check if the node is an internal pointer.
+ *
+ * \return
+ *    Return true if the node is an internal pointer, false otherwise.
+ *
+ * \note
+ *    This query currently only works on opcodes that can have a pinning array pointer,
+ *    i.e. aiadd, aiuadd, aladd, aluadd. Suppport for other opcodes can be added if necessary.
+ */
+bool
+OMR::Node::computeIsInternalPointer()
+   {
+   TR_ASSERT(self()->getOpCode().hasPinningArrayPointer(), "Opcode %s is not supported, node is " POINTER_PRINTF_FORMAT, self()->getOpCode().getName(), self());
+   return self()->computeIsCollectedReference();
+   }
+
+/**
+ * \brief
+ *    Check if the node is collected.
+ *
+ * \return
+ *    Return true if the node is collected, false otherwise.
+ *
+ * \note
+ *    This query traverses down the children of the node and try to find if the node originates
+ *    from something collected, i.e. anything that is collected reference or internal pointer.
+ */
+bool
+OMR::Node::computeIsCollectedReference()
+   {
+   TR::Node *curNode = self();
+   if (curNode->getOpCode().isTreeTop())
+      return false;
+
+   while (curNode)
+      {
+      if (!curNode->getType().isAddress())
+         return false;
+
+      if (curNode->isInternalPointer())
+         return true;
+
+      TR::ILOpCode op = curNode->getOpCode();
+      TR::ILOpCodes opValue = curNode->getOpCodeValue();
+
+      // If a language can handle a collected reference going via a non-address type,
+      // then the logic would need to be augmented to handle that
+      if (op.isConversion())
+         return false;
+
+      // The following are all opcodes that are address type, non-TreeTop and non-conversion
+      switch (opValue)
+         {
+         case TR::aiadd:
+         case TR::aiuadd:
+         case TR::aladd:
+         case TR::aluadd:
+            curNode = curNode->getFirstChild();
+            break;
+         // Symbols for calls and news does not contain collectedness information.
+         // Current implementation treats all object references collectable, and also
+         // assumes that the return of an acall* is an object reference.
+         case TR::acall:
+         case TR::acalli:
+         case TR::New:
+         case TR::newarray:
+         case TR::anewarray:
+         case TR::variableNew:
+         case TR::variableNewArray:
+         case TR::multianewarray:
+              return true;
+         // aload, aloadi, aRegLoad and loadaddr are associated with a symref, we should
+         // be able to tell its collectedness from its symref.
+         case TR::aRegLoad:
```

Comment:
Calling getSymbolReference on a register load (e.g. aregload) is a bit risky if the opcode does not return true for hasSymbolReference. That being said, I think there is at least one other spot in the compiler that accesses the symbol reference on a register load and I don't know of a specific reason why this would cause a functional problem as things stand.